### DISCUSSION TITLE
Flux CLI specific version

### DISCUSSION THREAD

**Discussion by greenpasta** [fluxcd/flux2#489] [Q&A] [answered]
I noticed that for the bootstrap you can install a specific version with `--version=x.y.z`. Is there any way to specify the version when installing the Flux CLI so it will match? Is this not much of a concern because the Flux CLI should be backwards compatible?

**Comment by stefanprodan** [comment, marked as answer]
To install Flux CLI you can use `brew install fluxcd/tap/flux`. To upgrade to the latest version `brew upgrade fluxcd/tap/flux`. You can find what version is installed with `flux -v` and use that for `flux bootstrap --version=v<CLI-VERSION>`.

It is recommended to upgrade the CLI before running bootstrap to upgrade the controllers with `flux bootstrap`. 

If you want to install an older version of Flux CLI, you can download the binary for your OS from the [release page](https://github.com/fluxcd/flux2/releases).

**Comment by debu99** [comment]
what if i don't want to install the latest client? how to do in brew install ?

**Reply by stefanprodan** [reply]
That's a good question to ask the Homebrew team. I don't think it's possible, there is a `brew extract --version` but seems broken to me.

**Reply by kingdonb** [reply]
Homebrew generally doesn't support installing other than the latest version, except in the case of language runtimes like Go or Python, where it is occasionally desirable to have multiple minor releases installed concurrently, and there is explicit support for this in the runtime.

You can for example `brew install go@1.15` but not `brew install go@1.15.14`, (you can only use brew to install the latest patch in a minor.)

**Comment by mousabaker** [comment]
With Bash you can run:

```
curl -s https://fluxcd.io/install.sh | sudo FLUX_VERSION=0.34.0 bash
```

**Reply by vishalathar** [reply]
`sudo curl -s https://fluxcd.io/install.sh | sudo FLUX_VERSION=0.34.0 bash` worked for me!

**Comment by rofreytag** [comment]
after `brew update` I now see `fluxcd/tap/flux@0.41` which is a relief. 

so you can `brew install fluxcd/tap/flux@0.41` which at least uses the recent major version.

Otherwise, brew would install (at the point of writing this post) a release candidate of `2.0.0-rc.3` which does not work correctly against clusters running flux `0.x` due to updates in the `CRDs`.

**Reply by rafaribe** [reply]
Yeah that's a bummer if you have multiple clusters with 2.0 and 0.x installed because either you need both versions of FluxCLI with different binaries. Would be awesome if Flux CLI was backwards compatible even if a major version release means that some things will be removed/deprecated.

**Reply by stefanprodan** [reply]
This will be the case after APIs graduate to GA. For example, GitRepositories and Kustomizations are now v1, all future Flux CLI versions will support them for years to come.

**Comment by asambhare** [comment]
my kubernetes clusters are running specific version of flux for bootstrap 2.0.0-rc.3 and on my machine I am using brew to install latest 2.0.0-rc.5 would it cause any issue or its recommended to have same version in both scenarios ?

**Reply by stefanprodan** [reply]
No issues with 2.0, all RCs and 2.0.x are compatible with each other. I recommend you upgrade Flux on your clusters to 2.0.1 as RCs had several bugs which were fixed.

**Reply by nachrivcost** [reply]
You can use the following tool to switch between Flux versions on your machine: https://github.com/nachrivcost/flxenv

**Comment by stefanprodan** [comment]
We now offer Homebrew formulas for all minor versions of Flux CLI.

To install the latest version:
```sh
brew install fluxcd/tap/flux
```

To install an older version specify the major.minor e.g.:
```sh
brew install fluxcd/tap/flux@2.2
```

There is also `flux@2.1`, `flux@2.0`, `flux@0.41`.

**Comment by dmeremyanin** [comment]
@stefanprodan since 2.4 is out, could you please release a formula for installing 2.3? Thanks

**Reply by stefanprodan** [reply]
`fluxcd/tap/flux@2.3` is now available
